File: ruby/version3_intl.rb
# http://localhost:4569/?n=10
# ruby version3_intl.rb

require 'sinatra'

set :port, 4569

get '/' do
  n = params['n']
  halt 400, 'Proporciona un número: http://localhost:4569/?n=10' if n.nil?

  content_type 'text/plain; charset=utf-8'
  convertir_espanol(n.to_i)
end

def convertir_espanol(n)
  return 'cero' if n == 0
  return 'menos ' + convertir_espanol(-n) if n < 0

  unidades = ['', 'uno', 'dos', 'tres', 'cuatro', 'cinco',
              'seis', 'siete', 'ocho', 'nueve', 'diez', 'once', 'doce',
              'trece', 'catorce', 'quince', 'dieciséis', 'diecisiete',
              'dieciocho', 'diecinueve', 'veinte']
  decenas = ['', '', 'veinte', 'treinta', 'cuarenta',
             'cincuenta', 'sesenta', 'setenta', 'ochenta', 'noventa']
  centenas = ['', 'ciento', 'doscientos', 'trescientos',
              'cuatrocientos', 'quinientos', 'seiscientos', 'setecientos',
              'ochocientos', 'novecientos']

  return unidades[n] if n <= 20
  return 'veinti' + unidades[n - 20] if n < 30

  if n < 100
    res = decenas[n / 10]
    res += ' y ' + unidades[n % 10] if n % 10 != 0
    return res
  end

  return 'cien' if n == 100

  if n < 1000
    res = centenas[n / 100]
    res += ' ' + convertir_espanol(n % 100) if n % 100 != 0
    return res
  end

  if n < 2000
    res = 'mil'
    res += ' ' + convertir_espanol(n % 1000) if n % 1000 != 0
    return res
  end

  if n < 1_000_000
    res = convertir_espanol(n / 1000) + ' mil'
    res += ' ' + convertir_espanol(n % 1000) if n % 1000 != 0
    return res
  end

  n.to_s
end

File: golang/version3_intl.go
// http://localhost:9092/?n=10
// go run version3_intl.go

package main

import (
	"fmt"
	"net/http"
	"strconv"
)

func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		n := r.URL.Query().Get("n")
		if n == "" {
			fmt.Fprint(w, "Proporciona un número: http://localhost:9092/?n=10")
			return
		}

		numero, err := strconv.ParseInt(n, 10, 64)
		if err != nil {
			http.Error(w, "Número inválido", 400)
			return
		}

		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
		fmt.Fprint(w, convertirEspanol(numero))
	})

	fmt.Println("Servidor en http://localhost:9092/?n=10")
	http.ListenAndServe(":9092", nil)
}

func convertirEspanol(n int64) string {
	if n == 0 {
		return "cero"
	}
	if n < 0 {
		return "menos " + convertirEspanol(-n)
	}

	unidades := []string{"", "uno", "dos", "tres", "cuatro", "cinco",
		"seis", "siete", "ocho", "nueve", "diez", "once", "doce",
		"trece", "catorce", "quince", "dieciséis", "diecisiete",
		"dieciocho", "diecinueve", "veinte"}
	decenas := []string{"", "", "veinte", "treinta", "cuarenta",
		"cincuenta", "sesenta", "setenta", "ochenta", "noventa"}
	centenas := []string{"", "ciento", "doscientos", "trescientos",
		"cuatrocientos", "quinientos", "seiscientos", "setecientos",
		"ochocientos", "novecientos"}

	switch {
	case n <= 20:
		return unidades[n]
	case n < 30:
		return "veinti" + unidades[n-20]
	case n < 100:
		res := decenas[n/10]
		if n%10 != 0 {
			res += " y " + unidades[n%10]
		}
		return res
	case n == 100:
		return "cien"
	case n < 1000:
		res := centenas[n/100]
		if n%100 != 0 {
			res += " " + convertirEspanol(n%100)
		}
		return res
	case n < 2000:
		res := "mil"
		if n%1000 != 0 {
			res += " " + convertirEspanol(n%1000)
		}
		return res
	case n < 1000000:
		res := convertirEspanol(n/1000) + " mil"
		if n%1000 != 0 {
			res += " " + convertirEspanol(n%1000)
		}
		return res
	default:
		return strconv.FormatInt(n, 10)
	}
}